File: launcher.py
#!/usr/bin/env python3
"""zenTTS Launcher — Rich TUI with proper box drawing and keyboard navigation."""

import os
import sys
import signal
import subprocess
import time
import termios
import tty
from pathlib import Path

try:
    from rich.console import Console
    from rich.panel import Panel
    from rich.table import Table
    from rich.text import Text
    from rich.align import Align
    from rich.layout import Layout
    from rich.live import Live
    from rich.style import Style
    from rich import box
except ImportError:
    print("Instalá rich: pip install rich")
    sys.exit(1)

PROJECT_DIR = Path(__file__).resolve().parent
PID_FILE = Path.home() / ".local/share/tts-zen/server.pid"
LOG_FILE = Path.home() / ".local/share/tts-zen/server.log"
PORT = 8765

PID_FILE.parent.mkdir(parents=True, exist_ok=True)

console = Console()
STYLE = {
    "purple": Style(color="#a78bfa"),
    "green": Style(color="#84cc76"),
    "red": Style(color="#f87171"),
    "yellow": Style(color="#fbbf24"),
    "gray": Style(color="#6b7280"),
    "white": Style(color="#f3f4f6", bold=True),
    "dim": Style(color="#6b7280", dim=True),
    "border": Style(color="#a78bfa"),
    "selected": Style(bgcolor="#7c3aed", color="#ffffff"),
    "key": Style(color="#a78bfa"),
}


def running() -> str | None:
    """Return PID if server is running."""
    try:
        result = subprocess.run(
            ["pgrep", "-f", "uvicorn tts_zen.main:app"],
            capture_output=True, text=True, timeout=2
        )
        if result.stdout.strip():
            return result.stdout.strip().split("\n")[0]
    except Exception:
        pass
    if PID_FILE.exists():
        try:
            pid = PID_FILE.read_text().strip()
            os.kill(int(pid), 0)
            return pid
        except Exception:
            pass
    return None


def server_healthy() -> bool:
    """Check if the server responds to health check."""
    try:
        result = subprocess.run(
            ["curl", "-s", f"http://127.0.0.1:{PORT}/health"],
            capture_output=True, text=True, timeout=3
        )
        return '"status":"ok"' in result.stdout
    except Exception:
        return False


def action_start():
    """Start the server in background."""
    if running():
        return
    env = os.environ.copy()
    server_dir = PROJECT_DIR / "server"
    subprocess.Popen(
        ["uv", "run", "uvicorn", "tts_zen.main:app", "--port", str(PORT), "--host", "127.0.0.1"],
        cwd=str(server_dir),
        stdout=open(LOG_FILE, "w"),
        stderr=subprocess.STDOUT,
        env=env,
        start_new_session=True,
    )
    # Wait for server to start
    for _ in range(20):
        time.sleep(0.3)
        if server_healthy():
            break
    # Send notification
    subprocess.run(
        ["notify-send", "-i", "audio-card", "zenTTS", f"Servidor listo en localhost:{PORT}"],
        capture_output=True, timeout=2
    )


def action_stop():
    """Stop the server."""
    try:
        subprocess.run(["pkill", "-f", "uvicorn tts_zen.main:app"], timeout=3)
    except Exception:
        pass
    time.sleep(0.5)
    PID_FILE.unlink(missing_ok=True)


def action_open():
    """Start server and open Zen Browser."""
    action_start()
    time.sleep(0.5)
    subprocess.Popen(
        ["flatpak", "run", "app.zen_browser.zen"],
        stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL,
        start_new_session=True,
    )


def build_ui(selected: int, extra: str = "") -> Layout:
    """Build the Rich layout."""
    layout = Layout()
    layout.split(
        Layout(name="header", size=5),
        Layout(name="status", size=7),
        Layout(name="menu", size=11),
    )

    # Header
    header_text = Text()
    header_text.append("\n zenTTS\n", STYLE["white"])
    header_text.append(" edge-tts · Zen Browser", STYLE["dim"])
    header_panel = Panel(
        Align.center(header_text),
        border_style=STYLE["border"],
        box=box.ROUNDED,
    )
    layout["header"].update(header_panel)

    # Status
    pid = running()
    healthy = server_healthy() if pid else False

    status_text = Text()
    if pid and healthy:
        status_text.append("  [●]  CORRIENDO  ", STYLE["green"])
        status_text.append(f"localhost:{PORT}\n", STYLE["dim"])
        status_text.append(f"  +  saludable    PID {pid}", STYLE["dim"])
    elif pid:
        status_text.append("  [●]  CORRIENDO  ", STYLE["yellow"])
        status_text.append(f"localhost:{PORT}\n", STYLE["dim"])
        status_text.append(f"  ~  sin respuesta    PID {pid}", STYLE["dim"])
    else:
        status_text.append("  [●]  DETENIDO  ", STYLE["red"])
        status_text.append("localhost:8765\n", STYLE["dim"])
        status_text.append("  -  servidor sin iniciar", STYLE["dim"])

    status_panel = Panel(
        status_text,
        title="Estado",
        border_style=STYLE["border"],
        box=box.ROUNDED,
        title_align="left",
    )
    layout["status"].update(status_panel)

    # Menu
    items = [
        "Iniciar servidor",
        "Detener servidor",
        "Refrescar estado",
        "Abrir Zen Browser",
    ]

    menu_text = Text()
    for i, item in enumerate(items):
        if i == selected:
            menu_text.append(f"    {item}", STYLE["selected"])
        else:
            menu_text.append(f"    {item}", STYLE["dim"])
        menu_text.append("\n")
    menu_text.append("\n")
    menu_text.append("  ^  v  mover    1-4 tecla    Enter elegir    q salir", STYLE["dim"])

    menu_panel = Panel(
        menu_text,
        title="Acciones",
        border_style=STYLE["border"],
        box=box.ROUNDED,
        title_align="left",
    )

    if extra:
        layout["menu"].update(Text(f"\n {extra}\n", style=STYLE["green"]))
        layout["menu"].update(menu_panel)
    else:
        layout["menu"].update(menu_panel)

    return layout


def get_key():
    """Read a single keypress including arrow keys."""
    fd = sys.stdin.fileno()
    old = termios.tcgetattr(fd)
    try:
        tty.setraw(fd)
        ch = sys.stdin.read(1)
        if ch == '\x1b':
            # Escape sequence
            ch2 = sys.stdin.read(1)
            if ch2 == '[':
                ch3 = sys.stdin.read(1)
                if ch3 == 'A': return 'up'
                if ch3 == 'B': return 'down'
                return ''
        return ch
    finally:
        termios.tcsetattr(fd, termios.TCSADRAIN, old)


def main():
    selected = 0
    extra = ""

    with Live(build_ui(selected), console=console, screen=True, refresh_per_second=10) as live:
        while True:
            live.update(build_ui(selected, extra))
            extra = ""

            key = get_key()
            if key in ('q', 'Q', '\x03'):  # \x03 = Ctrl+C
                break
            elif key == 'up':
                selected = (selected - 1) % 4
            elif key == 'down':
                selected = (selected + 1) % 4
            elif key in ('\r', '\n', ' '):  # Enter or Space
                if selected == 0:
                    extra = "  +  Iniciando servidor..."
                    live.update(build_ui(selected, extra))
                    action_start()
                    extra = "  +  Servidor listo" if running() else "  -  Error al iniciar"
                elif selected == 1:
                    extra = "  ~  Deteniendo..."
                    live.update(build_ui(selected, extra))
                    action_stop()
                    extra = "  +  Servidor detenido"
                elif selected == 2:
                    extra = "  ~  Refrescando..."
                elif selected == 3:
                    extra = "  ~  Abriendo Zen Browser..."
                    live.update(build_ui(selected, extra))
                    action_open()
                    extra = "  +  Zen Browser abierto"
            elif key == '1':
                selected = 0
            elif key == '2':
                selected = 1
            elif key == '3':
                selected = 2
            elif key == '4':
                selected = 3


if __name__ == "__main__":
    main()

File: launcher.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
# zenTTS Launcher — Ruby TUI with ANSI box drawing and keyboard navigation

require 'io/console'
require 'fileutils'

PORT = 8765
PROJECT_DIR = File.dirname(File.expand_path(__FILE__))
PID_DIR = File.join(Dir.home, '.local', 'share', 'tts-zen')
PID_FILE = File.join(PID_DIR, 'server.pid')
LOG_FILE = File.join(PID_DIR, 'server.log')
W = 46

FileUtils.mkdir_p(PID_DIR)

# ── ANSI ──
P = "\e[38;5;141m"; G = "\e[38;5;114m"; E = "\e[38;5;203m"
K = "\e[38;5;243m"; W = "\e[38;5;255m"; R = "\e[0m"
B = "\e[1m"; BS = "\e[48;5;99m"

def top()  "  #{P}╭#{'─' * (W-2)}╮#{R}" end
def bot()  "  #{P}╰#{'─' * (W-2)}╯#{R}" end
def mid(s) "  #{P}│#{R}#{s.ljust(W-2)}#{P}│#{R}" end

# ── Server helpers ──

def running?
  pid = `pgrep -f "uvicorn tts_zen.main:app" 2>/dev/null`.strip.split("\n").first
  return pid if pid && !pid.empty?
  if File.exist?(PID_FILE)
    pid = File.read(PID_FILE).strip
    return pid if pid && !pid.empty? && Process.kill(0, pid.to_i) rescue nil
  end
  nil
end

def healthy?
  r = `curl -s http://127.0.0.1:#{PORT}/health 2>/dev/null`
  r.include?('"status":"ok"')
end

def action_start
  return if running?
  pid = spawn(
    'ruby', 'server.rb',
    chdir: PROJECT_DIR,
    out: LOG_FILE, err: LOG_FILE,
    pgroup: true
  )
  Process.detach(pid)
  File.write(PID_FILE, pid.to_s)
  20.times do
    sleep 0.3
    break if healthy?
  end
  system('notify-send', '-i', 'audio-card', 'zenTTS', "localhost:#{PORT}", err: File::NULL) rescue nil
end

def action_stop
  system('pkill', '-f', 'uvicorn tts_zen.main:app', err: File::NULL) rescue nil
  sleep 0.5
  FileUtils.rm_f(PID_FILE)
end

def action_open
  action_start
  sleep 0.5
  spawn('flatpak', 'run', 'app.zen_browser.zen', out: File::NULL, err: File::NULL, pgroup: true)
end

# ── Drawing ──

def draw(sel, msg = '')
  pid = running?
  ok = pid ? healthy? : false

  if pid && ok
    status = "  #{G}#{B}CORRIENDO#{R}  #{K}localhost:#{PORT}#{R}  #{G}saludable#{R}  PID #{pid}"
  elsif pid
    status = "  #{G}#{B}CORRIENDO#{R}  #{K}localhost:#{PORT}#{R}  #{G}sin respuesta#{R}  PID #{pid}"
  else
    status = "  #{E}#{B}DETENIDO#{R}  #{K}localhost:#{PORT}#{R}  #{E}sin iniciar#{R}"
  end

  items = ['Iniciar servidor', 'Detener servidor', 'Refrescar estado', 'Abrir Zen Browser']
  lines = [
    '', top,
    mid("               #{W}#{B}zenTTS#{R}"),
    mid("        #{K}edge-tts · Zen Browser#{R}"),
    bot, '',
    top,
    mid("  #{B}Estado#{R}"),
    mid("  #{status}"),
    bot, ''
  ]

  lines << "  #{G}#{msg}#{R}" << '' unless msg.empty?

  lines += [
    top,
    mid("  #{B}Acciones#{R}")
  ]

  items.each_with_index do |item, i|
    if i == sel
      lines << mid("  #{BS}#{B}  #{item}  #{R}")
    else
      lines << mid("    #{K}#{item}#{R}")
    end
  end

  lines += [
    mid(''),
    mid("  #{K}^ v mover   1-4 tecla   Enter elegir   q salir#{R}"),
    bot
  ]

  $stdout.write("\e[2J\e[H" + lines.join("\n"))
  $stdout.flush
end

# ── Key reader ──

def read_key
  c = $stdin.getch
  if c == "\e"
    c2 = $stdin.getch
    if c2 == '['
      c3 = $stdin.getch
      return :up if c3 == 'A'
      return :down if c3 == 'B'
    end
    return ''
  end
  c
end

# ── Main ──

def main
  sel = 0
  msg = ''

  $stdout.write("\e[?25l\e[?1049h")
  $stdout.flush
  at_exit { $stdout.write("\e[?25h\e[?1049l"); $stdout.flush }

  loop do
    draw(sel, msg)
    msg = ''

    k = read_key
    break if ['q', 'Q', "\u0003"].include?(k)

    case k
    when :up   then sel = (sel - 1) % 4
    when :down then sel = (sel + 1) % 4
    when "\r", "\n", ' '
      case sel
      when 0 then msg = 'Iniciando...'; draw(sel, msg); action_start; msg = running? ? 'Servidor listo' : 'Error'
      when 1 then msg = 'Deteniendo...'; draw(sel, msg); action_stop; msg = 'Servidor detenido'
      when 2 then msg = 'Actualizado'
      when 3 then msg = 'Abriendo Zen...'; draw(sel, msg); action_open; msg = 'Zen Browser abierto'
      end
    when '1' then sel = 0
    when '2' then sel = 1
    when '3' then sel = 2
    when '4' then sel = 3
    end
  end
end

main
